ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/altair/statetransition/epoch/EpochProcessorAltair.java

```diff
@@ -67,8 +68,22 @@ public RewardAndPenaltyDeltas getRewardAndPenaltyDeltas(
     return calculator.getDeltas();
   }
 
+  /**
+   * Corresponds to process_participation_flag_updates in beacon-chain spec
+   *
+   * @param genericState The state to process
+   * @see <a
+   *     href="https://github.com/ethereum/eth2.0-specs/blob/master/specs/altair/beacon-chain.md#participation-flags-updates">Altair
+   *     Participation Flags updates</a>
+   */
   @Override
   public void processParticipationUpdates(final MutableBeaconState genericState) {
-    throw new NotImplementedException("TODO");
+    final MutableBeaconStateAltair state = MutableBeaconStateAltair.required(genericState);
+
+    state.getPreviousEpochParticipation().setAll(state.getCurrentEpochParticipation());
+
+    // Reset current epoch participation flags
+    state.getCurrentEpochParticipation().clear();
+    state.getCurrentEpochParticipation().setAll(SszByte.ZERO, state.getValidators().size());
```

Comment:
(nit) could be potentially optimized (if needed): 
```suggestion
    state.setPreviousEpochParticipation(state.getCurrentEpochParticipation());

    // Reset current epoch participation flags
    BeaconStateSchemaAltair schema = state.getSchema();
    state.setCurrentEpochParticipation(schema.getCurrentEpochParticipationSchema().getDefault());
    state.getCurrentEpochParticipation().setAll(SszByte.ZERO, state.getValidators().size());
```

Reply:
Curious - why is creating a default list cheaper than calling clear?

Reply:
Cheaper is `state.setPreviousEpochParticipation` (which needs to be added yet I believe) instead of setting all values one-by-one as implemented in `setAll()`. In other words the list is just _moved_ from `CurrentEpochParticipation` to `PreviousEpochParticipation` field. 

**But** after this operation both fields would have the reference to the same mutable list (see #3679) and thus 
```java
state.getCurrentEpochParticipation().clear();
```
would clear both fields. 
So you would need to _set_ a new list for `CurrentEpochParticipation` field. Then it's safe to modify it

Reply:
Creating default list and `clear()` should both be pretty fast

Reply:
But probably it's not worth doing any premature optimizations until they are appeared to be bottlenecks

Reply:
Oh, that makes more sense - I missed that part of the suggestion :D